Replace deprecated UIAlertView with UIAlertController

wutschel opened the pull request:
Replace deprecated `UIAlertView` with `UIAlertController`

FIxes https://github.com/xbmc/Official-Kodi-Remote-iOS/issues/198.

kambala-decapitator commented:
FYI I'm not going to add this to the upcoming update

wutschel commented:
I am fine with this. It will avoid risk for unwanted regressions, also I am still working on some of the other deprecated stuff like `UIActionSheet`. We can finalize this later.

kambala-decapitator commented:
please add UIActionSheet changes to this PR too

wutschel commented:
Now all traces of UIAlertView and UIActionSheet are gone.

wutschel commented:
Now also all traces of `UIPopoverController` are gone. Only very few "deprecated" warnings are left.

@kambala-decapitator, I am a bit lost in the deprecation warnings on `willAnimateRotation...`, `didRotateFrom...`, `willRotateTo...`, ` initWithRequest...` and `connectionWithRequest...`. You think you can take a look at this?

kambala-decapitator commented:
> @kambala-decapitator, I am a bit lost in the deprecation warnings on `willAnimateRotation...`, `didRotateFrom...`, `willRotateTo...`, ` initWithRequest...` and `connectionWithRequest...`. You think you can take a look at this?

hints in short if you'd like to explore:

- requests: `NSURLSession`
- rotation: `-viewWillTransitionToSize:withTransitionCoordinator:`

wutschel commented:
Rebased against master.

wutschel commented:
Rebased against master and added fix for longpress on filtered lists for iPad.

wutschel commented:
Rebased against master.

kambala-decapitator commented:
is this ready for review?

and what about [network calls and rotation](https://github.com/xbmc/Official-Kodi-Remote-iOS/pull/148#issuecomment-774539045) btw?

wutschel commented:
> is this ready for review?

Yes, it is.

> and what about [network calls and rotation](https://github.com/xbmc/Official-Kodi-Remote-iOS/pullt/148#issuecomment-774539045) btw?

I wasn´t able to rework this yet. Let us keep it like it is for now.

wutschel commented:
Rebase against master, no other changes.

wutschel commented:
@kambala-decapitator, you had a chance to look at this one? I know it´s quite some code changes, but would be good close on this one as we otherwise get into too many merge conflicts over time.

kambala-decapitator commented on XBMC Remote/NowPlaying.m:
```diff
@@ -1988,8 +1992,16 @@ -(void)showClearPlaylistAlert{
-        [alertView show];
+        UIAlertController *alertView = [UIAlertController alertControllerWithTitle:message message:nil preferredStyle:UIAlertControllerStyleAlert];
+        UIAlertAction* cancelButton = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", nil) style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {
+                // nothing to do
```
it's ok to skip such comment for ok/cancel buttons and just pass `nil` handler

kambala-decapitator commented on XBMC Remote/WebViewController.m:
```diff
@@ -91,14 +91,12 @@ - (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error {
-								   otherButtonTitles:nil];
-        [errorAlert show];
-//        [errorAlert release];
+        UIAlertController *alertView = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Error loading page", nil) message:nil preferredStyle:UIAlertControllerStyleAlert];
```
why did you remove error message?

wutschel replied:
Ouch, this was not intended. Re-added this locally.

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -319,4 +319,21 @@ + (CGRect)createCoverInsideJewel:(UIImageView*)jewelView jewelType:(eJewelType)t
+}
+
++ (UIAlertController*)createAlertCopyClipboard:(NSString*)title message:(NSString*)msg {
+    UIAlertController *alertView = [UIAlertController alertControllerWithTitle:title message:msg preferredStyle:UIAlertControllerStyleAlert];
```
I suggest to add a Cancel button as well, iirc there's an open issue about this

wutschel replied:
Agreed. Just added this locally.

kambala-decapitator commented on XBMC Remote/SettingsValuesViewController.m:
```diff
@@ -209,81 +209,83 @@ -(void)handleLongPress:(UILongPressGestureRecognizer *)gestureRecognizer {
+            break;
+        case cUnsupported:
+            return nil;
+            break;
```
useless statement

wutschel replied:
True, `break` is not reached in any case. Just kept it as I like to end `case` with `break`. Still removing the `break`?

kambala-decapitator replied:
Please remove

wutschel replied:
Ok, np.

kambala-decapitator commented on XBMC Remote/WebViewController.m:
```diff
@@ -105,15 +90,19 @@ -(IBAction)TwitterWebForwardButton:(id)sender{
+    [actionView addAction:cancelButton];
+    [actionView setModalPresentationStyle:UIModalPresentationPopover];
+    
+    UIPopoverPresentationController *popPresenter = [actionView popoverPresentationController];
```
you should check that it's not `nil` as the code below applies only to iPad

wutschel replied:
Just to make I got it right. You ask for the following (everywhere this segment occurs):

```
UIPopoverPresentationController *popPresenter = [actionView popoverPresentationController];
if (popPresenter != nil) {
    popPresenter.sourceView = bottomToolbar;
    popPresenter.sourceRect = bottomToolbar.bounds;
}
[self presentViewController:actionView animated:YES completion:nil];
```

kambala-decapitator replied:
Correct

wutschel replied:
Ok, applied for all relevant places where this was used. Will try to squash this into the commits.

kambala-decapitator commented on XBMC Remote/ViewControllerIPad.m:
```diff
@@ -304,35 +257,79 @@ - (void)actionSheet:(UIActionSheet *)actionSheet didDismissWithButtonIndex:(NSIn
+    [actionView addAction:cancelButton];
+    [actionView setModalPresentationStyle:UIModalPresentationPopover];
+    
+    UIPopoverPresentationController *popPresenter = [actionView popoverPresentationController];
```
`nil` check

kambala-decapitator commented on XBMC Remote/ViewControllerIPad.m:
```diff
@@ -304,35 +257,79 @@ - (void)actionSheet:(UIActionSheet *)actionSheet didDismissWithButtonIndex:(NSIn
+    jsonRPC = nil;
+    jsonRPC = [[DSJSONRPC alloc] initWithServiceEndpoint:[AppDelegate instance].getServerJSONEndPoint andHTTPHeaders:[AppDelegate instance].getServerHTTPHeaders];
+    [jsonRPC callMethod:action withParameters:params onCompletion:^(NSString *methodName, NSInteger callId, id methodResult, DSJSONRPCError *methodError, NSError* error) {
+        if (methodError==nil && error == nil){
+            UIAlertController *alertView = [Utilities createAlertOK:NSLocalizedString(@"Command executed", nil) message:nil];
+            [self presentViewController:alertView animated:YES completion:nil];
+        }
+        else{
+            UIAlertController *alertView = [Utilities createAlertOK:NSLocalizedString(@"Cannot do that", nil) message:nil];
+            [self presentViewController:alertView animated:YES completion:nil];
+        }
```
please put only title assignment inside if/else.

also just in case verify that completion is called on main thread.

kambala-decapitator commented on XBMC Remote/RemoteController.m:
```diff
@@ -676,61 +670,72 @@ - (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
+        [actionView setModalPresentationStyle:UIModalPresentationPopover];
+        
+        UIButton *subsButton = (UIButton *)[self.view viewWithTag:19];
+        UIPopoverPresentationController *popPresenter = [actionView popoverPresentationController];
```
nil check. Got tired of repeating this, please check all code :)

kambala-decapitator commented on XBMC Remote/RemoteController.m:
```diff
@@ -676,61 +670,72 @@ - (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
+        [actionView setModalPresentationStyle:UIModalPresentationPopover];
+        
+        UIButton *audioStreamsButton = (UIButton *)[self.view viewWithTag:20];
+        UIPopoverPresentationController *popPresenter = [actionView popoverPresentationController];
```
nil check

kambala-decapitator commented on XBMC Remote/DetailViewController.m:
```diff
@@ -3062,13 +3062,24 @@ -(IBAction)handleLongPress{
+                    showfromview = self.view;
+                }
+                else {
+                    showfromctrl = ([self doesShowSearchResults]) ? self.searchController : self;
```
drop brackets

wutschel replied:
Agreed, did this locally.

kambala-decapitator reviewed (changes requested):
I skipped couple of commits closer to the end because they're huge and not really readable, trust you :)

kambala-decapitator commented:
btw here's how to handle rotation: https://github.com/libsdl-org/SDL/blob/main/src/video/uikit/SDL_uikitviewcontroller.m#L336

wutschel commented on XBMC Remote/HostViewController.m:
```diff
@@ -367,9 +367,9 @@ -(void)netServiceDidResolveAddress:(NSNetService *)service {
-                NSURLConnection *connection = [NSURLConnection connectionWithRequest:request delegate:self];
-                [connection start];
+                NSURLSession *session = [NSURLSession sharedSession];
+                NSURLSessionDataTask *connection = [session dataTaskWithURL:url];
```
Interesting. Just checked and I can completely comment or remove the whole section

```
//                NSString *serverJSON=[NSString stringWithFormat:@"http://%@:%@/jsonrpc", ipUI.text, portUI.text];
//                NSURL *url = [[NSURL alloc] initWithString:serverJSON];
//                NSURLSession *session = [NSURLSession sharedSession];
//                NSURLSessionDataTask *connection = [session dataTaskWithURL:url];
//                [connection resume];
```

Is this just pinging the Kodi server and does not expect a response? The code is called when using "Add host" and "Find" to find Kodi instances in the local network.

kambala-decapitator replied:
Yes, if you need just to ping and don't care about response, then it's OK. 

But why would you want to remove it completely?

wutschel replied:
Will rename variables `pingSession` and `pingConnection`.

wutschel commented:
Added the required updates, rebased to master and squashed where reasonable. I could even combine some more commits (e.g. all UIAlertView changes into one single commit and all UIActionSheet into one single commit).

wutschel commented:
> btw here's how to handle rotation: https://github.com/libsdl-org/SDL/blob/main/src/video/uikit/SDL_uikitviewcontroller.m#L336

Thanks, could resolve this now.
